crates/miden-objects/src/asset/token_symbol.rs

```diff
@@ -53,68 +52,80 @@ impl TryFrom<Felt> for TokenSymbol {
 // Utils to encode and decode the token symbol as a Felt. Token Symbols can consists of up to 6
 // characters , e.g., A = 0, ...
-fn encode_symbol_to_felt(s: &str) -> Result<Felt, AssetError> {
+fn encode_symbol_to_felt(s: &str) -> Result<Felt, TokenSymbolError> {
     if s.is_empty() || s.len() > TokenSymbol::MAX_SYMBOL_LENGTH {
-        return Err(AssetError::TokenSymbolError(format!(
-            "token symbol of length {} is not between 1 and 6 characters long",
-            s.len()
-        )));
+        return Err(TokenSymbolError::InvalidLength(s.len()));
     } else if s.chars().any(|c| !c.is_ascii_uppercase()) {
-        return Err(AssetError::TokenSymbolError(format!(
-            "token symbol {} contains characters that are not uppercase ASCII",
-            s
-        )));
+        return Err(TokenSymbolError::InvalidCharacter(s.to_string()));
     }
 
     let mut encoded_value = 0;
     for char in s.chars() {
         let digit = char as u64 - b'A' as u64;
-        assert!(digit < 26);
-        encoded_value = encoded_value * 26 + digit;
+        assert!(digit < TokenSymbol::ALPHABET_LENGTH);
+        encoded_value = encoded_value * TokenSymbol::ALPHABET_LENGTH + digit;
     }
 
+    // add token length to the encoded value to be able to decode the exact number of characters
+    encoded_value = encoded_value * TokenSymbol::ALPHABET_LENGTH + s.len() as u64;
+
     Ok(Felt::new(encoded_value))
 }
 
-fn decode_felt_to_symbol(encoded_felt: Felt) -> String {
+fn decode_felt_to_symbol(encoded_felt: Felt) -> Result<String, TokenSymbolError> {
     let encoded_value = encoded_felt.as_int();
-    assert!(encoded_value < 26u64.pow(TokenSymbol::MAX_SYMBOL_LENGTH as u32));
+    assert!(encoded_value <= TokenSymbol::MAX_ENCODED_VALUE);
 
     let mut decoded_string = String::new();
     let mut remaining_value = encoded_value;
 
-    for _ in 0..6 {
-        let digit = (remaining_value % 26) as u8;
+    // get the token symbol length
+    let token_len = (remaining_value % TokenSymbol::ALPHABET_LENGTH) as usize;
+    if token_len == 0 || token_len > TokenSymbol::MAX_SYMBOL_LENGTH {
+        return Err(TokenSymbolError::InvalidLength(token_len));
+    }
+    remaining_value /= TokenSymbol::ALPHABET_LENGTH;
+
+    for _ in 0..token_len {
+        let digit = (remaining_value % TokenSymbol::ALPHABET_LENGTH) as u8;
         let char = (digit + b'A') as char;
         decoded_string.insert(0, char);
-        remaining_value /= 26;
+        remaining_value /= TokenSymbol::ALPHABET_LENGTH;
     }
 
-    decoded_string
+    // return an error if some data still remains after specified number of characters have been
+    // decoded.
+    if remaining_value != 0 {
+        return Err(TokenSymbolError::DataNotFullyDecoded);
+    }
+
+    Ok(decoded_string)
 }
 
 // TESTS
 // ================================================================================================
 #[test]
 fn test_token_symbol_decoding_encoding() {
```

Comment:
I would add another separate test for when the encoded length is less than the number of present characters. That should result in `DataNotFullyDecoded`.